jolt-core/src/r1cs/jolt_constraints.rs

```diff
@@ -0,0 +1,307 @@
+        cs.constrain_eq_conditional(JoltIn::OpFlags_IsLoad, packed_query, packed_load_store);
+        cs.constrain_eq_conditional(JoltIn::OpFlags_IsStore, packed_query, JoltIn::RAM_Read_RS2);
+
+        // TODO(sragss): Uses 2 excess constraints for condition gating. Could make constrain_pack_be_conditional... Or make everything conditional...
+        let chunked_x = cs.allocate_pack_be(
+            input_range!(JoltIn::ChunksX_0, JoltIn::ChunksX_3).to_vec(),
+            OPERAND_SIZE,
+        );
+        let chunked_y = cs.allocate_pack_be(
+            input_range!(JoltIn::ChunksY_0, JoltIn::ChunksY_3).to_vec(),
+            OPERAND_SIZE,
+        );
+        cs.constrain_eq_conditional(JoltIn::OpFlags_IsConcat, chunked_x, x);
+        cs.constrain_eq_conditional(JoltIn::OpFlags_IsConcat, chunked_y, y);
+
+        // if is_shift ? chunks_query[i] == zip(chunks_x[i], chunks_y[C-1]) : chunks_query[i] == zip(chunks_x[i], chunks_y[i])
+        let is_shift = JoltIn::IF_Sll + JoltIn::IF_Srl + JoltIn::IF_Sra;
+        let chunks_x = input_range!(JoltIn::ChunksX_0, JoltIn::ChunksX_3);
+        let chunks_y = input_range!(JoltIn::ChunksY_0, JoltIn::ChunksY_3);
+        let chunks_query = input_range!(JoltIn::ChunksQ_0, JoltIn::ChunksQ_3);
+        for i in 0..C {
+            let relevant_chunk_y =
+                cs.allocate_if_else(is_shift.clone(), chunks_y[C - 1], chunks_y[i]);
+            cs.constrain_eq_conditional(
+                JoltIn::OpFlags_IsConcat,
+                chunks_query[i],
+                (1i64 << 8) * chunks_x[i] + relevant_chunk_y,
+            );
+        }
+
+        // if (rd != 0 && update_rd_with_lookup_output == 1) constrain(rd_val == LookupOutput)
+        // if (rd != 0 && is_jump_instr == 1) constrain(rd_val == 4 * PC)
+        let rd_nonzero_and_lookup_to_rd =
+            cs.allocate_prod(JoltIn::Bytecode_RD, JoltIn::OpFlags_LookupOutToRd);
+        cs.constrain_eq_conditional(
+            rd_nonzero_and_lookup_to_rd,
+            JoltIn::RAM_Write_RD,
+            JoltIn::LookupOutput,
+        );
+        let rd_nonzero_and_jmp = cs.allocate_prod(JoltIn::Bytecode_RD, JoltIn::OpFlags_IsJmp);
+        let lhs = LC::sum2(JoltIn::PcIn, PC_START_ADDRESS - PC_NOOP_SHIFT);
+        let rhs = JoltIn::RAM_Write_RD;
+        cs.constrain_eq_conditional(rd_nonzero_and_jmp, lhs, rhs);
+
+        let branch_and_lookup_output =
+            cs.allocate_prod(JoltIn::OpFlags_IsBranch, JoltIn::LookupOutput);
+        let next_pc_jump = cs.allocate_if_else(
+            JoltIn::OpFlags_IsJmp,
+            JoltIn::LookupOutput + 4,
+            4 * JoltIn::PcIn + PC_START_ADDRESS + 4,
+        );
+
+        let next_pc_jump_branch = cs.allocate_if_else(
+            branch_and_lookup_output,
+            4 * JoltIn::PcIn + PC_START_ADDRESS + imm_signed,
+            next_pc_jump,
+        );
+        assert_static_aux_index!(next_pc_jump_branch, PC_BRANCH_AUX_INDEX);
```

Comment:
It constraints it to the constant `PC_BRANCH_AUX_INDEX` at runtime. This can then be used in `OffsetEqConstraints`.

TODO: Move the singular `OffsetEqConstraint` from `vm/mod.rs` into `jolt_constraints.rs`.